hal/rfcs/0000-spi-overhaul.md

```diff
@@ -0,0 +1,216 @@
+- Setting a frequency outside of the range given by `spi_get_capabilities()`.
+- Setting `bits` in `spi_format` to a value out of the range given by `spi_get_capabilities()`.
+- Passing an invalid pointer as `fill_symbol` to `spi_transfer` and `spi_transfer_async` while they would be required by the transfer (`rx_len != tx_len` or `tx==NULL`).
+- Passing an invalid pointer as `handler` to `spi_transfer_async`.
+- Calling `spi_transfer_async_abort()` while no async transfer is being processed (no transfer or a synchronous transfer).
+
+## Impact on partners' implementations
+
+The new API does not impact partners much as most of them factorise HAL implementation to family level.  
+For example :
+
+| Partner       | #of&nbsp;Implementation | details |
+| ---           | ---           | --- |
+| ST            | 1             | - |
+| Silicon Labs  | 1             | EFM32 family |
+| Analog device | 2             | They differ by ~20LoC that enable the ability to set the polarity and phase of the clock. They could most probably be merged as a single implementation. |
+| Nordic        | 2             | The first one for NRF51, the other for NRF52.<br>The functions calls are from what I have seen the same, this could be for Nordic a good opportunity to unify their SPI HAL implementation and bump NRF51’s SDK version. |
+| Atmel         | 2             | - one for the cortex-m4<br>- one for the cortex-m0+<br>This seems reasonable. |
+| Nuvoton       | 4             | implementation for 451, 472 and 480 are similar and look like the evolution of the 451.<br>NANO100's implementation has more "tweaks" but is really similar to the others, they could probably all be merged as 1. |
+| Maxim         | 6             | There's actually only 3 different implementations as :<br>- MAX32620<br>- MAX32600 = MAX32610<br>- MAX32630 = MAX32625 = MAX32620C |
+| NXP           | 13            | Most of them differ by the pin mapping that is provided in spi_api.c instead of `PeripheralPins.c` as it is done on most of other targets.<br>This number could be easily reduced to 2, maybe 4 in the very worst case. This would still be a huge maintenance gain for NXP DRY’ing their code base. |
+| Freescale     | 15            | Some of them are exactly the same. This number could be reduce to something around 5 implementation mostly due to the large number of µC generation (K20/KLXX/MCUXPresso etc). |
+
+The partner the most impacted by these changes is definitely Freescale/NXP but overall, in addition to the few bonus added by this new API, it would give partners' team a good opportunity to overhaul their implementations and factorise the code base.
+
+## Drawbacks
+--
+
+## Alternative solutions
+If it is decided to expose a peripheral interface instead of a "SPI channel to a slave", all `ss` management can be removed from hal and deferred to the driver (handling `ss` as any other GPIO).  
```

Comment:
Indeed, `ss` hw handling could be optional in master mode but needs to be used in slave mode.

This alternative solution is probably no longer relevant since `SPI_COUNT` and `spi_get_module` would now allow the C++ driver to determine which peripheral is in use.
We would then be able to have C++ driver creating 1 instance per slave while at HAL level we would only have 1 instance per peripheral.

Reply:
No drivers use the SPI API SS it because not all platforms support it, meaning it's effectively useless.

It's typically only implemented when the SPI peripheral supports it directly in hardware. If the HAL implementers were required to make the API work using internal GPIO if necessary, then drivers could use it.